Rationalize certain cases of texture usage flags

RandomShaper opened the pull request:
~The line above the one changed selects a format based on support for depth-stencil, but sampling was being included below among the required usages in addition to chose checked.~
**UPDATE:** Now this change is the complementary of the original one: instead of removing the usage flag, it's included in the checked usages.

- Check for exhaustive usage flags for depth buffer
  The current usage was fine with Vulkan, but other rendering APIs may not support sampling a 24-bit depth buffer and thus cause an error
- Remove uneeded storage flag from normal-roughness buffer
  Same here; other rendering APIs may not allow using MSAA textures as storage

clayjohn commented:
Isn't the sampling bit needed so that it can be read from a compute shader? e.g. when downsampling for screen-space effects like SSAO

RandomShaper commented:
If that happens, then yes. So, the fix should be the opposite: change the first line to check for sampling support. Would you agree?

RandomShaper committed:
Rationalize certain cases of texture usage flags

- Check for exhaustive usage flags for depth buffer
- Remove uneeded storage flag from normal-roughness buffer

RandomShaper commented on servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp:
```diff
@@ -435,7 +435,7 @@ void RenderForwardClustered::_allocate_normal_roughness_texture(RenderBufferData
 	tf.usage_bits = RD::TEXTURE_USAGE_SAMPLING_BIT | RD::TEXTURE_USAGE_STORAGE_BIT;
 
 	if (rb->msaa != RS::VIEWPORT_MSAA_DISABLED) {
-		tf.usage_bits |= RD::TEXTURE_USAGE_CAN_COPY_TO_BIT | RD::TEXTURE_USAGE_STORAGE_BIT;
```
This one is removed just because it's redundant.

clayjohn commented on servers/rendering/renderer_rd/forward_clustered/render_forward_clustered.cpp:
```diff
@@ -345,7 +345,7 @@ void RenderForwardClustered::RenderBufferDataForwardClustered::configure(RID p_c
 		color_msaa = RD::get_singleton()->texture_create(tf, RD::TextureView());
 
-		tf.format = RD::get_singleton()->texture_is_format_supported_for_usage(RD::DATA_FORMAT_D24_UNORM_S8_UINT, RD::TEXTURE_USAGE_DEPTH_STENCIL_ATTACHMENT_BIT) ? RD::DATA_FORMAT_D24_UNORM_S8_UINT : RD::DATA_FORMAT_D32_SFLOAT_S8_UINT;
+		tf.format = RD::get_singleton()->texture_is_format_supported_for_usage(RD::DATA_FORMAT_D24_UNORM_S8_UINT, RD::TEXTURE_USAGE_DEPTH_STENCIL_ATTACHMENT_BIT | RD::TEXTURE_USAGE_SAMPLING_BIT) ? RD::DATA_FORMAT_D24_UNORM_S8_UINT : RD::DATA_FORMAT_D32_SFLOAT_S8_UINT;
```
I don't understand this change. Don't we absolutely need ``TEXTURE_USAGE_DEPTH_STENCIL_ATTACHMENT_BIT``? To me this says that we can get away with using ``DATA_FORMAT_D24_UNORM_S8_UINT`` even if its unsupported as ``TEXTURE_USAGE_DEPTH_STENCIL_ATTACHMENT_BIT`` 

Perhaps you could explain in more detail?

RandomShaper replied:
The change means we are picking the format that, _in addition to_ being suited to a depth buffer, is also sampleable like a regular texture.

The latter is the piece I added. And it's needed since in D3D12 D24_S8 is an OK depth-stencil format but not sampleable, so we need the code to check for the full usage so it knows it has to pick D32_S8.

clayjohn reviewed (approved):
Looks good to me. I am not sure if @reduz wants to take a look at this. It is a small change where you are the subject matter expert. Lets just poke him and see if he responds.

akien-mga commented:
Thanks!
